raiden/utils/upgrades.py

```diff
@@ -0,0 +1,87 @@
+import os
+import shutil
+import sqlite3
+from pathlib import Path
+
+import sqlitebck
+import structlog
+
+from raiden.exceptions import RaidenDBUpgradeError
+from raiden.storage.serialize import JSONSerializer
+from raiden.storage.sqlite import RAIDEN_DB_VERSION, SQLiteStorage
+from raiden.storage.versions import older_db_file
+from raiden.utils.migrations.v16_to_v17 import upgrade_initiator_manager
+
+UPGRADES_LIST = [
+    upgrade_initiator_manager,
+]
+
+
+log = structlog.get_logger(__name__)
+
+
+class UpgradeManager:
+    """ This class is responsible for figuring out which migrations
+    need to be executed in order to bring the database up to date
+    with the current implementation.
+    """
+    def __init__(self, db_filename: str):
+        self._current_db_filename = Path(db_filename)
+
+    def run(self):
+        """
+        The `_db_filename` is going to hold the filename of the database
+        with the new version. However, the previous version's data
+        is going to exist in a file whose name contains the old version.
+        Therefore, running the migration means that we have to copy
+        all data to the current version's database, execute the migration
+        functions.
+        """
+        if self._current_db_filename.exists():
+            # The current version has already been created / updraded.
+            return
+
+        old_version, old_db_filename = older_db_file(str(self._current_db_filename.parent))
+
+        if not old_version:
+            # There are no older versions to upgrade from.
+            return
+
+        self._copy(str(old_db_filename), str(self._current_db_filename))
+
+        storage = SQLiteStorage(str(self._current_db_filename), JSONSerializer())
+
+        log.debug(f'Upgrading database to v{RAIDEN_DB_VERSION}')
+
+        try:
+            for upgrade_func in UPGRADES_LIST:
+                upgrade_func(storage, old_version, RAIDEN_DB_VERSION)
```

Comment:
This has to be executed inside a migration, to have an all or nothing upgrade, otherwise if the program crashes while the migrations are being applied, we will have a partial update.

The transaction must include the `update_version` to support the logic to check if the database is up-to-date

Reply:
Migrations are now executed in transactions as requested.